### PR TITLE
Add some Laplace-Beltrami-type operators

### PR DESCRIPTION
**Pull request by alexfikl**
Adds a Laplace-Beltrami, Yukawa-Beltrami and Helmholtz-Beltrami (not sure if this is their official name?) operator based on https://arxiv.org/abs/1705.00069.

There was some version of the Laplace-Beltrami one in `pytential/symbolic/old_diffop_primitives.py`, but this is hopefully cleaner and the interface is in line with the other scalar operators.

### CODE REVIEW

**Comment by alexfikl** [comment]
This seems to almost work as expected. For the Laplace-Beltrami operator it gives

h              | Error          | Running EOC                                                                                                                                                                       
:--------------|:---------------|:------------                                                                                                                                                                      
1.30152308e+00 | 7.35948067e-02 |                                                                                                                                                                                   
6.59334266e-01 | 1.76990591e-02 | 2.10                                                                                                                                                                              
3.30517323e-01 | 2.18716079e-03 | 3.03

but the Yukawa one doesn't seem to work. In the same setup it gives

h              | Error          | Running EOC                                                                                                                                                                       
:--------------|:---------------|:------------                                                                                                                                                                      
1.30152308e+00 | 1.74954628e-01 |                                                                                                                                                                                   
6.59334266e-01 | 7.86161956e-02 | 1.18                                                                                                                                                                              
3.30517323e-01 | 7.72197426e-02 | 0.03

This uses `target_order = qbx_order = 5` and the expected order is `3` because the formulations have two additional target derivatives in there. Not quite sure what it doesn't like for the Yukawa version. If it's a formulation issue, my guess would be that the two hypersingular kernels in there don't cancel as nicely (numerically) as in the case of Laplace.

File: pytential/symbolic/pde/beltrami.py [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+
+        # laplace
+        S0 = partial(sym.S, lknl, qbx_forced_limit=+1, **kwargs)        # noqa: N806
+        D0 = partial(sym.D, lknl, qbx_forced_limit="avg", **kwargs)     # noqa: N806
+        Dp0 = partial(sym.Dp, lknl, qbx_forced_limit="avg", **kwargs)   # noqa: N806
+
+        # base
+        S = partial(sym.S, knl, qbx_forced_limit=+1, **context)         # noqa: N806
+        Sp = partial(sym.Sp, knl, qbx_forced_limit="avg", **context)    # noqa: N806
+        Spp = partial(sym.Spp, knl, qbx_forced_limit="avg", **context)  # noqa: N806
```
Comment (alexfikl) [inline pytential/symbolic/pde/beltrami.py:114]: Are the `qbx_forced_limit` right here? `Sp` and `D` are probably right, but not sure about the hypersingular ones (here and below). Also for `S`, this only ever evaluates on the surface, so not quite sure if any side is to be prefered?

File: test/test_beltrami.py [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+    if visualize:
+        logging.basicConfig(level=logging.INFO)
+    actx = actx_factory()
+
+    radius = 1
+    case = eid.SphereTestCase(
+            target_order=5,
+            qbx_order=5,
+            source_ovsmp=8,
+            fmm_order=False, fmm_tol=None, fmm_backend=None,
```
Comment (alexfikl) [inline test/test_beltrami.py:129]: Tried turning on FMM here, but there's a lot of operators and composition of operators, so it cranks for a long time in code generation. Any preference? The test is marked as `slowtest`, so it won't have an effect on the CI.

File: pytential/symbolic/pde/beltrami.py (hunk 1 above)
Comment (inducer) [reply pytential/symbolic/pde/beltrami.py:162]: Yep, those look fine to me. `"avg"` gives the principal value/finite part, also for the hypersingular integrals.

File: pytential/symbolic/pde/beltrami.py [hunk 3]
```diff
@@ -0,0 +1,392 @@
...
+
+        # laplace
+        S0 = partial(sym.S, lknl, qbx_forced_limit=+1, **kwargs)        # noqa: N806
+        Sp0 = partial(sym.Sp, lknl, qbx_forced_limit="avg", **kwargs)   # noqa: N806
+        Dp0 = partial(sym.Dp, lknl, qbx_forced_limit="avg", **kwargs)   # noqa: N806
+
+        # base
+        S = partial(sym.S, knl, qbx_forced_limit=+1, **context)         # noqa: N806
+        Sp = partial(sym.Sp, knl, qbx_forced_limit="avg", **context)    # noqa: N806
+        Spp = partial(sym.Spp, knl, qbx_forced_limit="avg", **context)  # noqa: N806
```
Comment (inducer) [inline pytential/symbolic/pde/beltrami.py:205]: Avoid duplicating this code between above and here?

**Review by inducer — COMMENTED** [review]
Thanks for working on this! Good to go after a bit of code duplication is removed.

File: test/test_beltrami.py (hunk 2 above)
Comment (inducer) [reply test/test_beltrami.py]: Would any setting allow us to run this as part of CI? How long does it take?

File: pytential/symbolic/pde/beltrami.py (hunk 3 above)
Comment (alexfikl) [reply pytential/symbolic/pde/beltrami.py]: Went ahead and removed those `_get_left_operator` and `_get_right_operator` and unified this duplicated code in f922ecf.

**Comment by alexfikl** [comment]
I don't recall if the results from https://github.com/inducer/pytential/pull/115#issuecomment-933060755 for for 2D or 3D, but I re-ran the 2d test cases and they look very clean now.

For the Laplace-Beltrami on a circle it gives

h              | Left           | EOC         | Right          | EOC                                                                                                                                                                         
:--------------|:---------------|:------------|:---------------|:------------                                                                                                                                                                        
1.96349541e-01 | 2.02919448e-03 |             | 2.02920781e-03 |                                                                                                                                                                                     
9.81747704e-02 | 1.84161047e-04 | 3.46        | 1.84160989e-04 | 3.46                                                                                                                                                                                
6.54498469e-02 | 4.13140949e-05 | 3.69        | 4.13140834e-05 | 3.69                                                                                                                                                                                
4.90873852e-02 | 1.39383758e-05 | 3.78        | 1.39383719e-05 | 3.78                                                                                                                                                                                
Overall        |                | 3.59        |                | 3.59       

and for Yukawa-Beltrami on a circle it gives

h              | Left           | EOC         | Right          | EOC
:--------------|:---------------|:------------|:---------------|:------------
1.96349541e-01 | 1.97809390e-03 |             | 1.97810719e-03 |            
9.81747704e-02 | 1.80308226e-04 | 3.46        | 1.80308162e-04 | 3.46       
6.54498469e-02 | 4.05029809e-05 | 3.68        | 4.05029685e-05 | 3.68       
4.90873852e-02 | 1.36734037e-05 | 3.77        | 1.36733995e-05 | 3.77       
Overall        |                | 3.58        |                | 3.58

**Comment by inducer** [comment]
LGTM, thanks!

**Pull request merged**